central/auth/datastore/datastore_impl.go

```diff
@@ -46,24 +48,36 @@ func (d *datastoreImpl) listAuthM2MConfigsNoLock(ctx context.Context) ([]*storag
 }
 
 func (d *datastoreImpl) AddAuthM2MConfig(ctx context.Context, config *storage.AuthMachineToMachineConfig) (*storage.AuthMachineToMachineConfig, error) {
+	if err := sac.VerifyAuthzOK(accessSAC.WriteAllowed(ctx)); err != nil {
+		return nil, err
+	}
+
 	d.mutex.Lock()
 	defer d.mutex.Unlock()
+
+	ctx, tx, err := d.store.Begin(ctx)
+	if err != nil {
+		return nil, nil
+	}
+
 	exchanger, err := d.set.NewTokenExchangerFromConfig(ctx, config)
 	if err != nil {
-		return nil, err
+		return nil, d.wrapRollback(ctx, tx, err, config)
 	}
 
 	if err := d.store.Upsert(ctx, config); err != nil {
-		return nil, err
+		return nil, d.wrapRollback(ctx, tx, err, config)
 	}
 
-	if err := d.set.UpsertTokenExchanger(exchanger, config.GetIssuer()); err != nil {
-		return nil, err
-	}
-	return config, nil
+	d.set.UpsertTokenExchanger(exchanger, config.GetIssuer())
+	return config, tx.Commit(ctx)
```

Comment:
What will happen if this fails?

Reply:
@janisz raised a good point - perhaps we should first execute `tx.Commit()`, and attempt to rollback both in-memory and database. If there's no error, then we can execute `d.set.UpsertTokenExchanger(exchanger, config.GetIssuer())` as it does not produce any errors